manipulation/src/main/java/org/jboss/gm/manipulation/actions/UploadTaskTransformerAction.java

```diff
@@ -1,40 +1,96 @@
+import java.lang.reflect.Method;
+
 import org.gradle.api.Action;
 import org.gradle.api.Project;
-import org.gradle.api.artifacts.maven.MavenResolver;
-import org.gradle.api.tasks.Upload;
+import org.gradle.api.Task;
+import org.gradle.api.XmlProvider;
+import org.gradle.api.artifacts.ArtifactRepositoryContainer;
+import org.gradle.api.artifacts.repositories.ArtifactRepository;
+import org.gradle.api.logging.Logger;
+import org.gradle.util.GradleVersion;
+import org.jboss.gm.common.logging.GMLogger;
 import org.jboss.gm.common.model.ManipulationModel;
 import org.jboss.gm.manipulation.ResolvedDependenciesRepository;
 
 import static org.jboss.gm.manipulation.ManipulationPlugin.LEGACY_MAVEN_PLUGIN;
 
 /**
- * Fixes pom.xml generation in old "maven" plugin.
+ * Fixes {@code pom.xml} generation in the old &quot;maven&quot; plugin.
+ * <p>
+ * Adds {@code PomTransformer} to all {@code MavenResolver} repositories in {@code Upload} tasks.
  * <p>
- * Adds PomTransformer to all MavenResolver repositories in Upload tasks.
+ * <em>
+ * Gradle 7+ removed support for the old &quot;maven&quot; plugin. This class uses reflection in order to build and run
+ * on Gradle 7+. Due to the removal of the old &quot;maven&quot; plugin, this action does nothing when run on Gradle 7+.
+ * </em>
+ * </p>
  */
 public class UploadTaskTransformerAction implements Action<Project> {
 
-    private final ManipulationModel alignmentConfiguration;
-    private final ResolvedDependenciesRepository resolvedDependenciesRepository;
+    private static final Logger LOGGER = GMLogger.getLogger(UploadTaskTransformerAction.class);
+    private final boolean disabled;
+    private ManipulationModel alignmentConfiguration;
+    private ResolvedDependenciesRepository resolvedDependenciesRepository;
+    private Class<? extends ArtifactRepository> resolverClass;
+    private Method getPomMethod;
+    private Class<? extends Task> uploadClass;
+    private Method getRepositoriesMethod;
+    private Method withXmlMethod;
 
     public UploadTaskTransformerAction(ManipulationModel alignmentConfiguration,
             ResolvedDependenciesRepository resolvedDependenciesRepository) {
+        disabled = GradleVersion.current().compareTo(GradleVersion.version("7.0")) >= 0;
+
+        if (disabled) {
+            return;
+        }
+
         this.alignmentConfiguration = alignmentConfiguration;
         this.resolvedDependenciesRepository = resolvedDependenciesRepository;
+
+        try {
+            resolverClass = Class.forName("org.gradle.api.artifacts.maven.MavenResolver")
+                    .asSubclass(ArtifactRepository.class);
+            getPomMethod = resolverClass.getMethod("getPom");
+            // XXX: This class is in Gradle 7, but deprecated
+            // XXX: See <https://docs.gradle.org/current/userguide/upgrading_version_7.html#upload_task_deprecation>
+            uploadClass = Class.forName("org.gradle.api.tasks.Upload").asSubclass(Task.class);
```

Comment:
I'm slightly confused by this comment - if this class _is_ in Gradle 7 why is reflection required?

Reply:
I put a link to the manual which says that this class is deprecated and will be removed in Gradle 8. As it was using reflection for the other classes, it was just a bit of futureproofing.

Reply:
I guess it affects `LegacyMavenPublishingRepositoryAction` also, which uses this class directly. But, I guess you're saying not to worry about it until Gradle 8.

Reply:
Ok thanks, that makes sense.